SU2_CFD/src/solvers/CScalarSolver.cpp

```diff
@@ -0,0 +1,799 @@
+  SU2_OMP_MASTER {
+    /*--- Read the restart data from either an ASCII or binary SU2 file. ---*/
+
+    if (config->GetRead_Binary_Restart()) {
+      Read_SU2_Restart_Binary(geometry[MESH_0], config, restart_filename);
+    } else {
+      Read_SU2_Restart_ASCII(geometry[MESH_0], config, restart_filename);
+    }
+
+    /*--- Skip flow variables ---*/
+
+    unsigned short skipVars = nDim + solver[MESH_0][FLOW_SOL]->GetnVar();
+
+    /*--- Adjust the number of solution variables in the incompressible
+     restart. We always carry a space in nVar for the energy equation in the
+     mean flow solver, but we only write it to the restart if it is active.
+     Therefore, we must reduce skipVars here if energy is inactive so that
+     the turbulent variables are read correctly. ---*/
+
+    bool incompressible = (config->GetKind_Regime() == ENUM_REGIME::INCOMPRESSIBLE);
+    bool energy = config->GetEnergy_Equation();
+    bool weakly_coupled_heat = config->GetWeakly_Coupled_Heat();
+
+    if (incompressible && ((!energy) && (!weakly_coupled_heat))) skipVars--;
+
+    /*--- Load data from the restart into correct containers. ---*/
+
+    unsigned long counter = 0, iPoint_Global = 0;
+    for (; iPoint_Global < geometry[MESH_0]->GetGlobal_nPointDomain(); iPoint_Global++) {
+      /*--- Retrieve local index. If this node from the restart file lives
+       on the current processor, we will load and instantiate the vars. ---*/
+
+      auto iPoint_Local = geometry[MESH_0]->GetGlobal_to_Local_Point(iPoint_Global);
+
+      if (iPoint_Local > -1) {
+        /*--- We need to store this point's data, so jump to the correct
+         offset in the buffer of data from the restart file and load it. ---*/
+
+        index = counter * Restart_Vars[1] + skipVars;
+        for (iVar = 0; iVar < nVar; ++iVar) nodes->SetSolution(iPoint_Local, iVar, Restart_Data[index + iVar]);
+
+        /*--- Increment the overall counter for how many points have been loaded. ---*/
+        counter++;
+      }
+    }
+
+    /*--- Detect a wrong solution file ---*/
+
+    if (counter != nPointDomain) {
+      SU2_MPI::Error(string("The solution file ") + restart_filename + string(" doesn't match with the mesh file!\n") +
+                         string("It could be empty lines at the end of the file."),
+                     CURRENT_FUNCTION);
+    }
+
+  }  // end SU2_OMP_MASTER, pre and postprocessing are thread-safe.
+  END_SU2_OMP_MASTER
+  SU2_OMP_BARRIER
+
+  /*--- MPI solution and compute the eddy viscosity ---*/
+
+  solver[MESH_0][TURB_SOL]->InitiateComms(geometry[MESH_0], config, SOLUTION);
+  solver[MESH_0][TURB_SOL]->CompleteComms(geometry[MESH_0], config, SOLUTION);
+
+  solver[MESH_0][FLOW_SOL]->Preprocessing(geometry[MESH_0], solver[MESH_0], config, MESH_0, NO_RK_ITER,
+                                          RUNTIME_FLOW_SYS, false);
+  solver[MESH_0][TURB_SOL]->Postprocessing(geometry[MESH_0], solver[MESH_0], config, MESH_0);
```

Comment:
The other things that is solver specific part is especially what to Pre/Postprocess after solution loading to stay consistent with the 'normal' iteration. For species transport  with turb this gets a tiny bit more involved and with 4 solvers and more cross-dependencies one might have to do quite a bunch of calls to get everything updated.
(and the SOLVER_IDX needs to be specialized)